src/bundle.c

```diff
@@ -517,161 +501,222 @@ enum swupd_code remove_bundles(char **bundles)
+			char *dep;
+
+			/* the bundle is required by other bundles, make sure those bundles are not
+			 * also listed to be removed (the user may have selected those to be removed too) */
+			iter = list_head(simple_reqd_by);
 			while (iter) {
-				bundle = iter->data;
+				dep = iter->data;
+				if (list_search(bundles, dep, list_strcmp)) {
+					/* the dependency was requested to be removed too,
+					 * remove it from the list of dependencies */
+					iter = list_free_item(iter, free);
+					simple_reqd_by = iter;
+					continue;
+				}
 				iter = iter->next;
-				info("%s", bundle);
 			}
 
-			list_free_list_and_data(reqd_by, free);
-			ret = SWUPD_REQUIRED_BUNDLE_ERROR;
-			bad++;
-			goto out_free_mom;
-		}
+			/* do we still have dependencies? */
+			if (simple_reqd_by == NULL) {
+				goto no_deps;
+			}
 
-		current_mom->files = files_from_bundles(current_mom->submanifests);
-		current_mom->files = consolidate_files(current_mom->files);
+			/* the bundle is required by other bundles, do not continue with the removal
+			 * unless the --force flag is used */
+			if (!cmdline_option_force) {
+
+				/* let's try to build a dependency tree */
+				required_by(&tree_reqd_by, bundle, current_mom, 0, bundles, TREE_VIEW);
+				int number_of_tree_reqd = list_len(tree_reqd_by);
+
+				/* since we are considering exclusions (bundles that are dependencies
+				 * but were also added to the command for removal), we might have a
+				 * tree with no leafs here, if that is the case, do not print anything */
+				if (number_of_tree_reqd != 0) {
+
+					/* if the tree view is not too big we can display that in a nicer way,
+					 * if it exceeds the threshold, we will use the simplified list */
+					error("\n%s is required by the following bundles:\n", bundle);
+					if (number_of_tree_reqd > MAX_TREE) {
+						iter = list_head(simple_reqd_by);
+						while (iter) {
+							dep = iter->data;
+							iter = iter->next;
+							info(" - %s\n", dep);
+						}
+					} else {
+						iter = list_head(tree_reqd_by);
+						while (iter) {
+							dep = iter->data;
+							iter = iter->next;
+							info("%s", dep);
+						}
+						info("\nformat:\n");
+						info(" # * is-required-by\n");
+						info(" #   |-- is-required-by\n");
+						info(" # * is-also-required-by\n # ...\n");
+					}
+
+					info("\nBundle '%s' is required by %d bundle%s, skipping it...\n", bundle, number_of_reqd, number_of_reqd == 1 ? "" : "s");
+					info("Use \"swupd bundle-remove --force %s\" to remove '%s' and all bundles that require it\n", bundle, bundle);
+				}
 
-		/* Now that we have the consolidated list of all files, load bundle to be removed submanifest */
-		ret = load_bundle_manifest(bundle, subs, current_version, &bundle_manifest);
-		if (ret != 0 || !bundle_manifest) {
-			error("Cannot load %s sub-manifest (ret = %d)\n", bundle, ret);
-			bad++;
-			goto out_free_mom;
-		}
+				list_free_list_and_data(simple_reqd_by, free);
```

Comment:
Shouldn't we be freeing bundles_iter as well.

Reply:
bundles_iter is just a pointer to a list (which is pointed by bundles too), I'm using this pointer just so I don't loose the pointer during the while loop.... but as a second thought,I don't even need bundles_iter, I can just use a for loop instead of a while so I don't loose the pointer